tidy3d/plugins/invdes/postprocess.py

```diff
@@ -0,0 +1,131 @@
+    """Abstract base class defining components that make up postprocessing classes."""
+
+    @abc.abstractmethod
+    def evaluate(self, sim_data: td.SimulationData) -> float:
+        """How to evaluate this operation on a ``SimulationData`` object."""
+
+
+class CustomPostprocessOperation(AbstractPostprocessOperation):
+    """A postprocessing operation to subclass and implement ones own ``evalute`` method for."""
+
+    def evaluate(self, sim_data: td.SimulationData) -> float:
+        """How to evaluate this operation on a ``SimulationData`` object."""
+        raise NotImplementedError("Must define the 'self.evaluate(sim_data) -> float' method.")
+
+    @classmethod
+    def from_function(cls, fn: PostProcessFnType, **kwargs) -> CustomPostprocessOperation:
+        """Create a ``CustomPostprocessOperation`` from a function of ``SimulationData``."""
+        cls.evaluate = lambda self, sim_data: fn(sim_data)
+        obj = cls(**kwargs)
+        return obj
+
+
+class ElementaryPostprocessOperation(AbstractPostprocessOperation, abc.ABC):
+    """A postprocess operation that can work on its own."""
+
+
+class ModePower(ElementaryPostprocessOperation):
+    """Grab the power from a ``ModeMonitor`` and apply an optional weight."""
+
+    monitor_name: str = pd.Field(
+        ...,
+        title="Monitor Name",
+        description="Name of the ``ModeMonitor`` corresponding to the ``ModeData`` "
+        "that we want to compute power for.",
+    )
+
+    direction: td.components.types.Direction = pd.Field(
+        None,
+        title="Direction",
+        description="If specified, selects a specific direction ``'-'`` or ``'+'`` from the "
+        "``ModeData`` to include in power. Otherwise, sums over all directions.",
+    )
+
+    mode_index: pd.NonNegativeInt = pd.Field(
+        None,
+        title="Mode Index",
+        description="If specified, selects a specific mode index from the ``ModeData`` "
+        "to include in power. Otherwise, sums over all mode indices.",
+    )
+
+    f: float = pd.Field(
+        None,
+        title="Frequency",
+        description="If specified, selects a specific frequency from the ``ModeData`` "
+        "to include in power. Otherwise, sums over all frequencies.",
+    )
+
+    weight: float = pd.Field(
+        1.0,
+        title="Weight",
+        description="Weight specifying the contribution of this power to the objective function. ",
+    )
+
+    @property
+    def sel_kwargs(self) -> dict[str, typing.Any]:
+        """Selection kwargs corresponding to the fields."""
+        sel_kwargs_all = dict(direction=self.direction, mode_index=self.mode_index, f=self.f)
+        return {key: sel for key, sel in sel_kwargs_all.items() if sel is not None}
+
+    def evaluate(self, sim_data: td.SimulationData) -> float:
+        """Evaluate this instance when passed a simulation dataset."""
+
+        mnt_data = sim_data[self.monitor_name]
```

Comment:
Do we need to check first if that monitor exists or is that always guaranteed?

Reply:
This will throw a `KeyError` [here](https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/base_sim/data/sim_data.py#L46).

But I'll add a check just to make things a bit more user friendly.